File: Enigma.Net/Enigma/BusinessLogic.cs
using System.Collections.Generic;
using System.IO;

namespace Enigma
{
    public class BusinessLogic
    {
        private List<Roll> Rolls { get; set; }
        private List<Roll> RollsReverse { get; set; }

        public BusinessLogic(List<Roll>rolls)
        {
            Rolls = rolls;

            RollsReverse = new List<Roll>();
            RollsReverse.AddRange(rolls);
            RollsReverse.Reverse();
        }

        public void TransformFile(string inputFfilename, string outputFilename, Enums.Mode mode)
        {
            const int buffersize = 65536;

            using (FileStream fileInStream = new FileStream(inputFfilename, FileMode.Open, FileAccess.Read))
            {
                using (FileStream fileOutStream = new FileStream(outputFilename, FileMode.Create, FileAccess.Write))
                {
                    byte[] buffer = new byte[buffersize];
                    int readCount = 0;
                    while ((readCount = fileInStream.Read(buffer, 0, buffersize)) > 0)
                    {
                        TransformByteArray(buffer, mode);
                        fileOutStream.Write(buffer, 0, readCount);
                    }
                }
            }
        }

        public void TransformByteArray(byte[] input, Enums.Mode mode)
        {
            for (int i = 0; i < input.Length; i++)
            {
                if (mode == Enums.Mode.Enc)
                    foreach (Roll roll in Rolls)
                        input[i] = roll.Enc(input[i]);

                else if (mode == Enums.Mode.Dec)
                    foreach (Roll roll in RollsReverse)
                        input[i] = roll.Dec(input[i]);

                RollOn();
            }
        }

        private void RollOn()
        {
            foreach(Roll roll in Rolls)
            {
                if(!roll.RollOn())
                    break;
            }
        }
    }
}

File: Enigma.Java/src/main/BusinessLogic.java
package main;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BusinessLogic {
	private List<Roll> rolls;
	private List<Roll> rollsReverse;

	public BusinessLogic(List<Roll> rolls) {

		this.rolls = rolls;

		rollsReverse = new ArrayList<Roll>(rolls);
		Collections.reverse(rollsReverse);
	}

	public void TransformFile(String inputFfilename, String outputFilename, Enums.Mode mode) throws IOException {
		final int buffersize = 65536;

		try (

				FileInputStream fileInStream = new FileInputStream(inputFfilename);
				FileOutputStream fileOutStream = new FileOutputStream(outputFilename);) {

			byte[] buffer = new byte[buffersize];
			int readCount = 0;

			while ((readCount = fileInStream.read(buffer, 0, buffersize)) > 0) {
				TransformByteArray(buffer, mode);
				fileOutStream.write(buffer, 0, readCount);
			}
		}
	}

	public void TransformByteArray(byte[] input, Enums.Mode mode) {

		if (mode == Enums.Mode.Encode) {

			for (int i = 0; i < input.length; i++) {
				for (Roll roll : rolls)
					input[i] = roll.Encrypt(input[i]);

				RollOn();
			}
		}

		if (mode == Enums.Mode.Decode) {
			for (int i = 0; i < input.length; i++) {
				for (Roll roll : rollsReverse)
					input[i] = roll.Decrypt(input[i]);

				RollOn();
			}
		}
	}

	private void RollOn() {
		for (Roll roll : rolls) {
			if (!roll.RollOn())
				break;
		}
	}
}
